src/measure_theory/function/ae_eq_of_integral.lean

```diff
@@ -0,0 +1,486 @@
+  {t : set α} (ht : measurable_set t) (hμt : μ t ≠ ∞) :
+  0 ≤ᵐ[μ.restrict t] f :=
+begin
+  refine ae_nonneg_restrict_of_forall_set_integral_nonneg_inter hμt
+    (hf_int_finite t ht (lt_top_iff_ne_top.mpr hμt)) (λ s hs, _),
+  refine (hf_zero (s ∩ t) (hs.inter ht) _),
+  exact (measure_mono (set.inter_subset_right s t)).trans_lt (lt_top_iff_ne_top.mpr hμt),
+end
+
+lemma ae_eq_zero_restrict_of_forall_set_integral_eq_zero_real {f : α → ℝ}
+  (hf_int_finite : ∀ s, measurable_set s → μ s < ∞ → integrable_on f s μ)
+  (hf_zero : ∀ s, measurable_set s → μ s < ∞ → ∫ x in s, f x ∂μ = 0)
+  {t : set α} (ht : measurable_set t) (hμt : μ t ≠ ∞) :
+  f =ᵐ[μ.restrict t] 0 :=
+begin
+  suffices h_and : f ≤ᵐ[μ.restrict t] 0 ∧ 0 ≤ᵐ[μ.restrict t] f,
+    from h_and.1.mp (h_and.2.mono (λ x hx1 hx2, le_antisymm hx2 hx1)),
+  refine ⟨_, ae_nonneg_restrict_of_forall_set_integral_nonneg hf_int_finite
+    (λ s hs hμs, (hf_zero s hs hμs).symm.le) ht hμt⟩,
+  suffices h_neg : 0 ≤ᵐ[μ.restrict t] -f,
+  { refine h_neg.mono (λ x hx, _),
+    rw pi.neg_apply at hx,
+    simpa using hx, },
+  refine ae_nonneg_restrict_of_forall_set_integral_nonneg
+    (λ s hs hμs, (hf_int_finite s hs hμs).neg) (λ s hs hμs, _) ht hμt,
+  simp_rw pi.neg_apply,
+  rw [integral_neg, neg_nonneg],
+  exact (hf_zero s hs hμs).le,
+end
+
+end real
+
+lemma ae_eq_zero_restrict_of_forall_set_integral_eq_zero_R_or_C {f : α → 𝕜}
+  (hf_int_finite : ∀ s, measurable_set s → μ s < ∞ → integrable_on f s μ)
+  (hf_zero : ∀ s, measurable_set s → μ s < ∞ → ∫ x in s, f x ∂μ = 0)
+  {t : set α} (ht : measurable_set t) (hμt : μ t ≠ ∞) :
+  f =ᵐ[μ.restrict t] 0 :=
+begin
+  suffices h_re_im : (∀ᵐ x ∂(μ.restrict t), is_R_or_C.re (f x) = 0)
+    ∧ ∀ᵐ x ∂(μ.restrict t), is_R_or_C.im (f x) = 0,
+  { rw ← eventually_and at h_re_im,
+    refine h_re_im.mono (λ x hx, _),
+    rwa [is_R_or_C.ext_iff, pi.zero_apply, add_monoid_hom.map_zero, add_monoid_hom.map_zero], },
+  have hf_re : ∀ s, measurable_set s → μ s < ∞ → integrable_on (λ x, is_R_or_C.re (f x)) s μ,
+    from λ s hs hμs, (hf_int_finite s hs hμs).re,
+  have hf_im : ∀ s, measurable_set s → μ s < ∞ → integrable_on (λ x, is_R_or_C.im (f x)) s μ,
+    from λ s hs hμs, (hf_int_finite s hs hμs).im,
+  have hf_zero_re : ∀ s, measurable_set s → μ s < ∞ → ∫ x in s, is_R_or_C.re (f x) ∂μ = 0,
+  { intros s hs hμs,
+    rw [integral_re (hf_int_finite s hs hμs), hf_zero s hs hμs, is_R_or_C.zero_re'], },
+  have hf_zero_im : ∀ s, measurable_set s → μ s < ∞ → ∫ x in s, is_R_or_C.im (f x) ∂μ = 0,
+  { intros s hs hμs,
+    rw [integral_im (hf_int_finite s hs hμs), hf_zero s hs hμs],
+    simp only [add_monoid_hom.map_zero], },
+  exact ⟨ae_eq_zero_restrict_of_forall_set_integral_eq_zero_real hf_re hf_zero_re ht hμt,
+    ae_eq_zero_restrict_of_forall_set_integral_eq_zero_real hf_im hf_zero_im ht hμt⟩,
+end
+
+variables [is_scalar_tower ℝ 𝕜 E]
+include 𝕜
+
+lemma ae_eq_zero_restrict_of_forall_set_integral_eq_zero {f : α → E}
```

Comment:
as you mention in the TODO, it feels strange to have this lemma only for inner product spaces over \bbk, while it holds true for any second countable real vector space. Do you prefer that we merge the PR in its current state and generalize later, or would you consider this for the current PR?

Reply:
I would rather leave it as it is now, restricted to inner product spaces, and come back to it later. There are other parts of the measure_theory folder which suffer from the same issue (l2_space). I can either extend all of it to real vector spaces or continue with the artificial restriction for now.
I'd like to get to the end of the conditional expectation definition sooner rather than later, because I have already spent a lot of time going down rabbit holes (integrability conditions, finiteness conditions on the measure leading to strongly measurable functions...) and I feel like I should instead get it done, even if it is not in the greatest possible generality.

Reply:
That is completely reasonable!

Reply:
In fact, all the lemmas for Banach spaces were already available, so I did the generalization.